### PR TITLE
Make fe_field_function compute::point locations use the gridtools version

### PR DESCRIPTION
**Pull request by GivAlz**
After creating the function compute point locations I'm making the one in FEFieldFunction use the one in GridTools.

This should bring a performance boost, also to many other functions. The only caveat: there might be problems with curved meshes which are very deformed, as reported in the documentation of GridTools::compute_point_locations.

Best,
Giovanni

### CODE REVIEW

File: include/deal.II/numerics/fe_field_function.templates.h [hunk 1]
```diff
@@ -438,162 +442,19 @@ namespace Functions
...
-        qps += qpoints[n].size();
-      }
-    Assert(qps == np,
-           ExcDimensionMismatch(qps, np));
-#endif
-
-    return c;
+    // Getting the reference to the underlying triangulation and initializing the cache
+    const Triangulation<dim,DoFHandlerType::space_dimension> &tria = dh->get_triangulation();
+    GridTools::Cache<dim,DoFHandlerType::space_dimension> cache(tria,mapping);
```
Comment (luca-heltai) [inline include/deal.II/numerics/fe_field_function.templates.h]: Can you add a member `cache` to the `FEFieldFunction` class, initialized at construction time? In this way you don't recompute the cache at each round.

**Comment by GivAlz** [comment]
Cache added to FEFieldFunction and now works/the test fe_compute_point_location passes :)

**Comment by GivAlz** [comment]
Tests are failing, ok... I'll try rebasing but probably it's because some cells in the output are repeated.
Anyway I probably can solve the problem, I shall update tomorrow :)

**Comment by GivAlz** [comment]
I opened the other PR because tests related to fe_field_function are failing giving additional output/giving out messages such as dealii::GridTools::ExcPointNotFound<2> and not timeout. I believe it is because sometimes we get twice the same cell in the output vector of fe_field_function::grid_tools (the diameter is discarding points which actually are in the cell). Once the other PR is merged I think we shall discover quickly if this was the case. 

Anyway I don't know well deal.II tests so I might be completely wrong, if so please just tell me why you think these tests gave the wrong output.

Best,
Giovanni

The following tests FAILED:
	 84 - numerics/generalized_interpolation_03.debug (Failed)
	124 - numerics/generalized_interpolation_02.debug (Failed)
	139 - numerics/generalized_interpolation_04.debug (Failed)
	160 - numerics/generalized_interpolation_05.debug (Failed)
	2583 - bits/fe_field_function_03.debug (Failed)
	2714 - bits/fe_field_function_01.debug (Failed)
	2840 - bits/fe_field_function_01_vector.debug (Failed)

**Comment by luca-heltai** [comment]
Please rebase. 3 of the failing tests should now pass. Hopefully also the other 4 have been resolved.

**Comment by GivAlz** [comment]
In an attempt to find the error I modified again the code, adding the possibility for a hint cell in compute_point_locations (and a test for this...). This part is ok.

Then I noticed all failing tests gave an error in line 1684 of grid tools (the tests failing are in fact using fe_field_function): this is the function find active cell around point failing. I tried to see if I was updating the cache correctly, but everything seemed ok. In the end I tried with the following (inspired from the failing tests):

`Triangulation<2> triangulation;
GridGenerator::hyper_cube(triangulation, -0.3, 0.7);
triangulation.refine_global(2 == 2 ? 2 : 1);
MappingQ<2> mapping(1);
GridTools::Cache<2,2> cache(triangulation,mapping);
GridTools::find_active_cell_around_point (cache, Point<2>(-0.3,-0.3));`

It fails, also for `GridTools::find_active_cell_around_point (cache, Point<2>(0,-0.3));`: I believe there's some bug with find active cell around point when the points are on the external edge of the mesh...

**Comment by bangerth** [comment]
If I read the comment trail correctly, then this is currently blocked by #5512. Let us know once you want us to take another look.

**Comment by GivAlz** [comment]
I belive it's important that interpolating tests, which make use of the compute_cell_around_points function, pass. So yes, @bangerth, I believe that once #5512 is merged and I rebase, everything should do just fine(ok, maybe something will still time out). If that's not the case I shall ask for some more help ;) Thank you!

**Comment by GivAlz** [comment]
Finally tester is happy :)

**Comment by luca-heltai** [comment]
Can you rebase?

**Pull request merged**